File: app/src/main/java/com/bitflaker/lucidsourcekit/main/goals/GoalStatisticsCalculator.java
package com.bitflaker.lucidsourcekit.main.goals;

import com.bitflaker.lucidsourcekit.database.MainDatabase;
import com.bitflaker.lucidsourcekit.database.goals.entities.resulttables.DetailedShuffleHasGoal;
import com.bitflaker.lucidsourcekit.utils.Tools;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Single;

public class GoalStatisticsCalculator {
    private int count;
    private int achieved;
    private float difficulty;
    private float recurrenceFrequency;
    private float shuffleOccurrenceRating;
    private List<DetailedShuffleHasGoal> goals;
    private final long timestampDayStart;
    private final long timestampDayEnd;
    private final MainDatabase db;
    private final boolean isDayDone;

    public GoalStatisticsCalculator(MainDatabase db, long timestampDayStart, long timestampDayEnd) {
        this.db = db;
        this.timestampDayStart = timestampDayStart;
        this.timestampDayEnd = timestampDayEnd;
        this.isDayDone = Tools.isTimeInPast(timestampDayEnd);
        this.count = 0;
        this.achieved = 0;
        this.difficulty = 0.0f;
        this.recurrenceFrequency = 0.0f;
        this.shuffleOccurrenceRating = 0.0f;
    }

    public Single<GoalStatisticsCalculator> calculate() {
        return Single.fromCallable(() -> {
            float totalAverageDifficulty = db.getGoalDao().getAverageDifficulty().blockingGet().floatValue();

            List<Integer> goalIds = new ArrayList<>();
            Maybe<List<DetailedShuffleHasGoal>> maybeGoalsCurrent = db.getShuffleHasGoalDao().getShuffleFrom(timestampDayStart, timestampDayEnd);
            if(!maybeGoalsCurrent.isEmpty().blockingGet()) {
                goals = maybeGoalsCurrent.blockingGet();
                if(goals.isEmpty()) {
                    return this;
                }
                int countAchieved = 0;
                float difficultySum = 0;
                for (DetailedShuffleHasGoal goal : goals) {
                    if(isDayDone && goal.achieved) {
                        countAchieved++;
                    }
                    difficultySum += goal.difficulty;
                    goalIds.add(goal.goalId);
                }
                achieved = countAchieved;
                count = goalIds.size();
                difficulty = difficultySum / count;
                shuffleOccurrenceRating = difficulty / totalAverageDifficulty;
                int totalDrawCount = db.getShuffleHasGoalDao().getCountOfGoalsDrawn(goalIds, timestampDayStart, timestampDayEnd).blockingGet();
                int totalCountOfGoalsShuffled = db.getShuffleHasGoalDao().getAmountOfTotalDrawnGoals(timestampDayStart, timestampDayEnd).blockingGet();
                recurrenceFrequency = 100 * totalDrawCount / (float) totalCountOfGoalsShuffled;
            }

            return this;
        });
    }

    public int getCount() {
        return count;
    }

    public int getAchieved() {
        return achieved;
    }

    public float getDifficulty() {
        return difficulty;
    }

    public float getRecurrenceFrequency() {
        return recurrenceFrequency;
    }

    public float getShuffleOccurrenceRating() {
        return shuffleOccurrenceRating;
    }

    public List<DetailedShuffleHasGoal> getGoals() {
        return goals;
    }

    public float getRatioAchieved() {
        return count == 0 ? 0 : achieved / (float) count;
    }

    public boolean isDayDone() {
        return isDayDone;
    }

    public boolean hasGoals() {
        return count > 0;
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/main/goals/GoalStatisticsCalculator.kt
package com.bitflaker.lucidsourcekit.main.goals

import com.bitflaker.lucidsourcekit.database.MainDatabase
import com.bitflaker.lucidsourcekit.database.goals.entities.Goal
import com.bitflaker.lucidsourcekit.database.goals.entities.resulttables.DetailedShuffleHasGoal
import com.bitflaker.lucidsourcekit.utils.Tools

class GoalStatisticsCalculator(
    private val db: MainDatabase,
    private val dayStartTimestamp: Long,
    private val dayEndTimestamp: Long
) {
    private var count: Int = 0
    private var achieved: Int = 0
    private var goals: List<DetailedShuffleHasGoal>? = null
    private val isDayDone: Boolean = Tools.isTimeInPast(dayEndTimestamp)
    var difficulty: Float = 0.0f
        private set
    var recurrenceFrequency: Float = 0.0f
        private set
    var shuffleOccurrenceRating: Float = 0.0f
        private set

    suspend fun calculate(): GoalStatisticsCalculator {
        val totalAverageDifficulty = db.goalDao.getAverageDifficulty()
        val goalIds = ArrayList<Int>()
        val currentGoals = db.shuffleHasGoalDao.getShuffleFrom(dayStartTimestamp, dayEndTimestamp)

        if (currentGoals.isEmpty()) {
            return this
        }

        goals = currentGoals
        var countAchieved = 0
        var difficultySum = 0f
        for (goal in currentGoals) {
            if (isDayDone && false /* TODO: REMOVE OLD GOAL ACHIEVED APPROACH */) {
                countAchieved++
            }
            difficultySum += goal.difficulty
            goalIds.add(goal.goalId)
        }
        achieved = countAchieved
        count = goalIds.size
        difficulty = difficultySum / count
        shuffleOccurrenceRating = (difficulty / totalAverageDifficulty).toFloat()
        val totalDrawCount: Int = db.shuffleHasGoalDao.getCountOfGoalsDrawn(goalIds, dayStartTimestamp, dayEndTimestamp)
        val totalCountOfGoalsShuffled: Int = db.shuffleHasGoalDao.getAmountOfTotalDrawnGoals(dayStartTimestamp, dayEndTimestamp)
        recurrenceFrequency = 100 * totalDrawCount / totalCountOfGoalsShuffled.toFloat()

        return this
    }

    fun getGoals(): List<Goal>? {
        return goals?.map {
            Goal(
                it.goalId,
                it.description,
                it.difficulty,
                it.difficultyLocked
            )
        }
    }

    val shuffleId: Int
        get() {
            val currentGoals = goals
            if (currentGoals.isNullOrEmpty()) {
                return 0
            }
            return currentGoals[0].shuffleId
        }

    val ratioAchieved: Float
        get() = if (count == 0) 0f else achieved / count.toFloat()

    fun hasGoals(): Boolean {
        return count > 0
    }
}
